File: oracle/tokenize_abi.py
from __future__ import annotations

import re
from typing import Any, Dict, Iterable, Optional, Tuple


def _line_col(text: str, idx: int) -> Tuple[int, int]:
    line = text.count("\n", 0, idx) + 1
    col = idx - text.rfind("\n", 0, idx)
    return line, col


def _char_in_ranges(ch: str, ranges: Iterable[Tuple[int, int]]) -> bool:
    cp = ord(ch)
    for start, end in ranges:
        if start <= cp <= end:
            return True
    return False


def abi_tokenize_ok(text: str, tokenizer_abi: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    allowed_ranges = tokenizer_abi.get("allowed_unicode_ranges")
    disallowed_ranges = tokenizer_abi.get("disallowed_unicode_ranges")
    allowed_regex = tokenizer_abi.get("allowed_char_regex")
    disallowed_regex = tokenizer_abi.get("disallowed_char_regex")

    allowed_pattern = re.compile(allowed_regex) if allowed_regex else None
    disallowed_pattern = re.compile(disallowed_regex) if disallowed_regex else None

    for idx, ch in enumerate(text):
        if disallowed_ranges and _char_in_ranges(ch, disallowed_ranges):
            line, col = _line_col(text, idx)
            return {
                "code": "E_TOK_DISALLOWED_CHAR",
                "msg": f"disallowed character U+{ord(ch):04X}",
                "line": line,
                "col": col,
            }
        if allowed_ranges and not _char_in_ranges(ch, allowed_ranges):
            line, col = _line_col(text, idx)
            return {
                "code": "E_TOK_OUT_OF_RANGE",
                "msg": f"character U+{ord(ch):04X} outside allowed ranges",
                "line": line,
                "col": col,
            }
        if allowed_pattern and not allowed_pattern.fullmatch(ch):
            line, col = _line_col(text, idx)
            return {
                "code": "E_TOK_REGEX_MISMATCH",
                "msg": f"character U+{ord(ch):04X} failed allowed regex",
                "line": line,
                "col": col,
            }
        if disallowed_pattern and disallowed_pattern.search(ch):
            line, col = _line_col(text, idx)
            return {
                "code": "E_TOK_DISALLOWED_REGEX",
                "msg": f"character U+{ord(ch):04X} matched disallowed regex",
                "line": line,
                "col": col,
            }

    return None

File: oracle-js/tokenize_abi.js
function charInRanges(ch, ranges) {
  const cp = ch.codePointAt(0);
  return ranges.some(([start, end]) => cp >= start && cp <= end);
}

export function abiTokenizeOk(text, tokenizerAbi) {
  const allowedRanges = tokenizerAbi?.allowed_unicode_ranges;
  const disallowedRanges = tokenizerAbi?.disallowed_unicode_ranges;
  const allowedRegex = tokenizerAbi?.allowed_char_regex
    ? new RegExp(tokenizerAbi.allowed_char_regex)
    : null;
  const disallowedRegex = tokenizerAbi?.disallowed_char_regex
    ? new RegExp(tokenizerAbi.disallowed_char_regex)
    : null;

  for (let i = 0; i < text.length; i += 1) {
    const ch = text[i];
    if (disallowedRanges && charInRanges(ch, disallowedRanges)) {
      return { code: "E_TOK_DISALLOWED_CHAR", msg: `disallowed U+${ch.codePointAt(0).toString(16).padStart(4, "0")}` };
    }
    if (allowedRanges && !charInRanges(ch, allowedRanges)) {
      return { code: "E_TOK_OUT_OF_RANGE", msg: `U+${ch.codePointAt(0).toString(16).padStart(4, "0")} outside allowed ranges` };
    }
    if (allowedRegex && !allowedRegex.test(ch)) {
      return { code: "E_TOK_REGEX_MISMATCH", msg: `U+${ch.codePointAt(0).toString(16).padStart(4, "0")} failed allowed regex` };
    }
    if (disallowedRegex && disallowedRegex.test(ch)) {
      return { code: "E_TOK_DISALLOWED_REGEX", msg: `U+${ch.codePointAt(0).toString(16).padStart(4, "0")} matched disallowed regex` };
    }
  }

  return null;
}
